### ISSUE TITLE
Callbacks in generic tensorflow trainer

### ISSUE THREAD

**Issue by arnaudvl** [SeldonIO/alibi-detect#9]
Basic functionality to cover:

- [ ] Early stopping
- [ ] Learning rate scheduling

**Comment by mishinma** [comment]
Hi @arnaudvl. What's the status of this issue? Do you plan to integrate with tensorboard at some point?

A side question: why is it necessary to have a custom trainer (vs keras model.fit)?

**Comment by arnaudvl** [comment]
Haven't been able to start on this issue yet unfortunately. Adding tensorboard support should definitely be done too. I'll open a separate issue for that.

I find that it allows easier fine grained control over training. It is also more straightforward to add more complex custom losses (e.g. the AEGMM one which takes the multiple outputs from the model as input or the adversarial VAE one using an additional external model in the loss) as simple functions without the need to change the internals of the model (e.g. using ```self.add_loss()``` in the forward pass). The reason why I want to avoid this if possible is that it removes some flexibility for the end user who might want to change loss functions dependent on their use case. I am also working on some sequence outlier detection models with training strategies (e.g. mix between teacher forcing and using previous model output as input) that are relatively easy to add to the custom trainer but not so much with ```.fit()```.  Of course the downside is that adding in callbacks etc. requires additional work. Hope this makes it more clear!

**Comment by mishinma** [comment]
Thank you for your answer! That makes sense to me.

**Comment by arnaudvl** [comment]
Would be dealt with by #114 .
